Add legible tags issue 331.

alexgaganashvili opened the pull request.

Commit:
add function to create additional tag while copying between registries

Commit:
add function to create additional tag while copying between registries

joaopapereira commented on pkg/imgpkg/util/tag.go:
```diff
@@ -28,3 +30,20 @@ func BuildDefaultUploadTagRef(item WithDigest, importRepo regname.Repository) (r
+	// ex.:
+	// 1. registry.f.q.d.n/myorg/my-img@sha256:012345679e2a06317dc02abcde123759eb9e6105a0731224a0c63898abcde123
+	// 2. registry.f.q.d.n/myorg/my-img:some-tag@sha256:012345679e2a06317dc02abcde123759eb9e6105a0731224a0c63898abcde123
+	origRepoAndTag := strings.Split(itemRef, "@")
```
This can be accomplished with 
```go
digest, err := regname.NewDigest(itemRef)

origRepoPath := a.RepositoryStr()
dashedRepo := strings.ReplaceAll(origRepoPath, "/", "-")
tag := fmt.Sprintf("%s-%s-%s.imgpkg", dashedRepo[len(dashedRepo)-49:], digest.Algorithm, digest.Hex)
```

I did not try to compile this, so please take it with a grain of salt, also there is the problem of `dashedRepo` being smaller than 49

joaopapereira reviewed (changes requested):
The issue was changed to accommodate a different syntax, do you mind updating your PR to match the new requirements?

Commit:
create additional tag if ---use-repo-based-tags flag is specified

Commit:
resolve conflicts

Commit:
resolve conflicts in pkg/imgpkg/internal/util/tag.go

joaopapereira commented on pkg/imgpkg/imageset/image_set.go:
```diff
@@ -31,7 +31,7 @@ func NewImageSet(concurrency int, ui goui.UI) ImageSet {
 
 func (i ImageSet) Relocate(foundImages *UnprocessedImageRefs,
-	importRepo regname.Repository, registry registry.ImagesReaderWriter) (*ProcessedImages, error) {
+	importRepo regname.Repository, registry registry.ImagesReaderWriter, useRepoBasedTags bool) (*ProcessedImages, error) {
```
Instead of the Relocate to take a boolean flag I personally would prefer to have

```go
type TagGenerator interface {
   Generate(sourceDigest regname.Digest, destinationRepo regname.Repository) (regname.Tag, error)
}

func NewImageSet(concurrency int, ui goui.UI, tagGen TagGenerator) ImageSet {
	return ImageSet{concurrency, UI, tagGen}
}
```

This would allow us to have eventually other types of tag generation in the future if we want and the decision of a particular tag type is not made deep inside the copy code but is done by the caller of ImageSet

joaopapereira commented on pkg/imgpkg/imageset/image_set.go:
```diff
@@ -146,27 +148,40 @@ func checkForAnyAsyncErrors(imgOrIndexes []imagedesc.ImageOrIndex, errCh chan er
-	switch {
-	case item.Image != nil:
-		artifactToWrite, err = i.mountableImage(*item.Image, uploadTagRef, registry)
+	tagArr := make([]regname.Tag, 1, 2)
```
I think we talked about only having 1 tag, so we do not need a slice here

alexgaganashvili replied:
Got it. That's because of my comment: https://github.com/vmware-tanzu/carvel-imgpkg/issues/331#issuecomment-1046136094

joaopapereira commented on pkg/imgpkg/imageset/image_set.go:
```diff
@@ -146,27 +148,40 @@ func checkForAnyAsyncErrors(imgOrIndexes []imagedesc.ImageOrIndex, errCh chan er
-	switch {
-	case item.Image != nil:
-		artifactToWrite, err = i.mountableImage(*item.Image, uploadTagRef, registry)
+	tagArr := make([]regname.Tag, 1, 2)
```
I think we talked about only having 1 tag, so we do not need a slice here

joaopapereira commented on pkg/imgpkg/internal/util/tag.go:
```diff
@@ -29,3 +30,27 @@ func BuildDefaultUploadTagRef(item WithDigest, importRepo regname.Repository) (r
 }
+
+// BuildLegibleUploadTagRef Builds a legible tag from original image reference
+func BuildLegibleUploadTagRef(item WithDigest, itemRef string, importRepo regname.Repository) (regname.Tag, error) {
```
I was looking at this code and I am not 100% sure this works in the scenario:
1. copy image from registry A to registry B
2. copy image from registry B to registry C

Looking at the parameters that are provided to this function I believe that `itemRef` on the second copy will be the reference from registry B. 

What we need is the original location of the Image. This might require more changes in the `imagedesc` package to save the original location of the Image as part of the ImageDescriptor type.

Can you validate if the case I pointed out in this comment is really happening?
Let me know if this comment is a bit to much information and might confuse you, I am free to jump on a Zoom call and look into this issue with you if it helps

joaopapereira replied:
I was looking at this code and I am not 100% sure this works in the scenario:
1. copy image from registry A to registry B
2. copy image from registry B to registry C

Looking at the parameters that are provided to this function I believe that `itemRef` on the second copy will be the reference from registry B. 

What we need is the original location of the Image. This might require more changes in the `imagedesc` package to save the original location of the Image as part of the ImageDescriptor type.

Can you validate if the case I pointed out in this comment is really happening?
Let me know if this comment is a bit to much information and might confuse you, I am free to jump on a Zoom call and look into this issue with you if it helps

joaopapereira reviewed (changes requested):
I added some comments on this PR. Not 100% sure if this implementation will work on all cases. Let me know if you need more information about my comments.

joaopapereira committed:
Add flag to do not retrieve cosign artifacts while describe bundle (#345)

Signed-off-by: Joao Pereira <joaod@vmware.com>

joaopapereira committed:
Apimachinery bump (#343)

* Make the diff check more specific

Signed-off-by: Joao Pereira <joaod@vmware.com>

* Bump apimachinery to 0.23.4

Signed-off-by: Joao Pereira <joaod@vmware.com>

joaopapereira committed:
Images Location not correctly created

When Nested Bundles are not in the same repository as the Outer Bundle,
imgpkg was incorrectly relying on the Full Described Name of the Bundle
to find if it is a Bundle or not.

This only affected copy from tar to repository

Signed-off-by: Joao Pereira <joaod@vmware.com>

benmoss committed:
Add linux-arm64 to releases

joaopapereira committed:
Update cobra to version 1.4.0

Signed-off-by: Joao Pereira <joaod@vmware.com>

Commit:
create tag generators dynamically; with --use-repo-based-tags, generate/preserve tags based on the value image in ImagesLock

fix lint errors

fix lint errors

fix lint errors

add copyright header

long repo name not passing external registry tests

fix env var-s for executing tests on github

fix: add missing IMGPKG_E2E_RELOCATION_REPO_C value for tests in external registries

fixing value of RelocationRepo2

fix message in copy command usage

Commit:
create tag generators dynamically; with --use-repo-based-tags, generate/preserve tags based on the value image in ImagesLock

Commit:
fix lint errors

Commit:
fix lint errors

Commit:
fix lint errors

Commit:
add copyright header

Commit:
long repo name not passing external registry tests

Commit:
fix env var-s for executing tests on github

Commit:
fix: add missing IMGPKG_E2E_RELOCATION_REPO_C value for tests in external registries

Commit:
fixing value of RelocationRepo2

Commit:
fix message in copy command usage

Commit:
add event to pull_request_target types

Commit:
add event to pull_request_target types

Commit:
Merge branch 'develop' into add-legible-tags-issue-331

Commit:
remove Relocationrepo2

Commit:
remove Relocationrepo2

cppforlife commented on pkg/imgpkg/cmd/copy.go:
```diff
@@ -74,6 +84,8 @@ func NewCopyCmd(o *CopyOptions) *cobra.Command {
 	cmd.Flags().IntVar(&o.Concurrency, "concurrency", 5, "Concurrency")
 	cmd.Flags().BoolVar(&o.IncludeNonDistributable, "include-non-distributable-layers", false,
 		"Include non-distributable layers when copying an image/bundle")
+	cmd.Flags().BoolVar(&o.UseRepoBasedTags, "use-repo-based-tags", false,
```
use-repo-based-tags => repo-based-tags (to be more consistent with other carvel tool flag naming)

cppforlife commented on pkg/imgpkg/internal/util/tag.go:
```diff
@@ -5,16 +5,103 @@ package util
+}
+
+// NewTagGenerator returns different TagGenerator types
+func NewTagGenerator(tagGenType string) TagGenerator {
```
instead of introducing this function lets just instantiate RepoBasedTagGenerator or DefaultTagGenerator in call sites.

cppforlife commented on test/helpers/env.go:
```diff
@@ -13,15 +13,16 @@ import (
+	Image           string
+	ImgpkgPath      string
+	RelocationRepo  string
+	RelocationRepo2 string
```
what is RelocationRepo2?

alexgaganashvili replied:
Forgot to remove that one from env.go. I don't need it. Thx.

Commit:
resolve review comments

Commit:
resolve review comments

joaopapereira reviewed (approved):
LGTM

joaopapereira commented on pkg/imgpkg/imagedesc/types.go:
```diff
@@ -21,8 +21,9 @@ type ImageOrIndex struct {
 	Labels map[string]string
-}
 
+	OrigRef string
```
```suggestion
	OrigRef string `json:"-"`
```

joaopapereira replied:
Ignore this

joaopapereira commented on pkg/imgpkg/imagedesc/types.go:
```diff
@@ -57,6 +58,7 @@ type ImageIndexDescriptor struct {
 	Digest    string
 	Raw       string
 	Tag       string
+	OrigRef   string
```
```suggestion
	OrigRef   string `json:"-"`
```

joaopapereira replied:
Ignore this

joaopapereira reviewed (changes requested):
I made these 2 suggestions after talking with @cppforlife. By doing this we ensure that the same tar created with an old version of imgpkg and a version with this feature will have the same `sha256sum`

joaopapereira reviewed (approved):
LGTM

Commit:
Merge branch 'add-legible-tags-issue-331' of github.com:alexgaganashvili/carvel-imgpkg into add-legible-tags-issue-331

Commit:
make suggested changes
